lms/djangoapps/lti_provider/signature_validator.py

```diff
@@ -0,0 +1,182 @@
+"""
+
+from django.core.exceptions import ObjectDoesNotExist
+
+from oauthlib.oauth1 import SignatureOnlyEndpoint
+from oauthlib.oauth1 import RequestValidator
+
+from lti_provider.models import LtiConsumer
+
+
+class SignatureValidator(RequestValidator):
+    """
+    Helper class that verifies the OAuth signature on a request.
+
+    The pattern required by the oauthlib library mandates that subclasses of
+    RequestValidator contain instance methods that can be called back into in
+    order to fetch the consumer secret or to check that fields conform to
+    application-specific requirements.
+    """
+
+    def __init__(self):
+        super(SignatureValidator, self).__init__()
+        self.endpoint = SignatureOnlyEndpoint(self)
+
+    # The OAuth signature uses the endpoint URL as part of the request to be
+    # hashed. By default, the oauthlib library rejects any URLs that do not
+    # use HTTPS. We turn this behavior off in order to allow edX to run without
+    # SSL in development mode. When the platform is deployed and running with
+    # SSL enabled, the URL passed to the signature verifier must start with
+    # 'https', otherwise the message signature would not match the one generated
+    # on the platform.
+    enforce_ssl = False
+
+    def check_client_key(self, key):
+        """
+        Verify that the key supplied by the LTI consumer is valid for an LTI
+        launch. This method is only concerned with the structure of the key;
+        whether the key is associated with a known LTI consumer is checked in
+        validate_client_key. This method signature is required by the oauthlib
+        library.
+
+        :return: True if the client key is valid, or False if it is not.
+        """
+        return key is not None and 0 < len(key) <= 32
+
+    def check_nonce(self, nonce):
+        """
+        Verify that the nonce value that accompanies the OAuth signature is
+        valid. This method is concerned only with the structure of the nonce;
+        the validate_timestamp_and_nonce method will check that the nonce has
+        not been used within the specified time frame. This method signature is
+        required by the oauthlib library.
+
+        :return: True if the OAuth nonce is valid, or False if it is not.
+        """
+        return nonce is not None and 0 < len(nonce) <= 64
+
+    def validate_timestamp_and_nonce(self, client_key, timestamp, nonce,
+                                     request, request_token=None,
+                                     access_token=None):
+        """
+        Verify that the request is not too old (according to the timestamp), and
+        that the nonce value has not been used already within the period of time
+        in which the timestamp marks a request as valid. This method signature
+        is required by the oauthlib library.
+
+        :return: True if the OAuth nonce and timestamp are valid, False if they
+        are not.
+        """
+        return True
```

Comment:
In this particular case we can't, since the underlying OAuth implementation uses this method as part of the signature workflow. If it returns anything but true, it flags the signature as invalid.

That said, there are a lot of other methods that have to be implemented in this class (since they are marked as abstract in the superclass), and NotImplementedError probably makes more sense in those than 'pass'.